searx/query.py

```diff
@@ -63,100 +70,225 @@ def _parse_query(self):
+
+            if (value == lang_id
+                or value == lang_name
+                or value == english_name
+                or value.replace('-', ' ') == country)\
+               and value not in self.languages:
+                found = True
+                lang_parts = lang_id.split('-')
+                if len(lang_parts) == 2:
+                    self.languages.append(lang_parts[0] + '-' + lang_parts[1].upper())
+                else:
+                    self.languages.append(lang_id)
+                # to ensure best match (first match is not necessarily the best one)
+                if value == lang_id:
+                    break
+
+        # user may set a valid, yet not selectable language
+        if VALID_LANGUAGE_CODE.match(value):
+            lang_parts = value.split('-')
+            if len(lang_parts) > 1:
+                value = lang_parts[0].lower() + '-' + lang_parts[1].upper()
+            if value not in self.languages:
+                self.languages.append(value)
+                found = True
+
+        if found:
+            return True
+        elif enable_autocomplete:
+            self._language_autocomplete(value)
+            return False
+
+    def _external_bang_autocomplete(self):
+        self.autocomplete_list.append('!!g')
+        self.autocomplete_list.append('!!ddg')
+        self.autocomplete_list.append('!!bing')
+
+    def _external_bang_parse(self, value, enable_autocomplete):
+        found = False
+        bang_definition, bang_ac_list = get_bang_definition_and_autocomplete(value)
+        if bang_definition is not None:
+            self.external_bang = value
+            found = True
+
+        if enable_autocomplete:
+            for bang_autocomplete in bang_ac_list:
+                self.autocomplete_list.append('!!' + bang_autocomplete)
+        return found
+
+    def _bang_autocomplete(self, first_char, value):
+        if len(value) == 0:
+            # show some example queries
+            for suggestion in ['images', 'wikipedia', 'osm']:
+                if suggestion not in self.disabled_engines or suggestion in categories:
+                    self.autocomplete_list.append(first_char + suggestion)
+        else:
+            # check if query starts with categorie name
+            for categorie in categories:
+                if categorie.startswith(value):
+                    self.autocomplete_list.append(first_char + categorie)
+
+            # check if query starts with engine name
+            for engine in engines:
+                if engine.startswith(value):
+                    self.autocomplete_list.append(first_char + engine.replace(' ', '_'))
+
+            # check if query starts with engine shortcut
+            for engine_shortcut in engine_shortcuts:
+                if engine_shortcut.startswith(value):
+                    self.autocomplete_list.append(first_char + engine_shortcut)
+
+    def _bang_parse(self, first_char, value, enable_autocomplete):
```

Comment:
Please extract `_bang_autocomplete` in a similar fashion as in `_language_parse`.

Reply:
[`get_bang_definition_and_autocomplete`](https://github.com/searx/searx/blob/e2fb50089219207f7e11bcfc049b48f2f4b9a6c8/searx/external_bang.py#L47) returns the bang_definition and the autocompletion information (since the external bangs are stored in a trie, it is "free").

Now `ExternalBangParser._parse` returns `found, bang_ac_list` but doesn't add anything to `autocomplete_list`.